### ISSUE TITLE
module get stucks and all subsequent passive items are not responsive

### ISSUE THREAD

**Issue by ptempier** [monitoringartist/zabbix-docker-monitoring#73]
The last responsive line before the receiving thread is stuck is always the same: "Docker's socket query"
At about the same times i get a "Cannot open metric file"
When all threads are exhausted, they are all in processing states and passive items dont works anymore.
This is agent version 3.2.6
```
juin 20 23:32:48 aaaa zabbix_agentd[6747]: __zbx_zbx_setproctitle() title:'listener #3 [processing request]'
juin 20 23:32:48 aaaa zabbix_agentd[6747]: Requested [docker.stats[/ffff,memory_stats,usage]]
juin 20 23:32:48 aaaa zabbix_agentd[6747]: In zbx_module_docker_stats()
juin 20 23:32:48 aaaa zabbix_agentd[6747]: In zbx_module_docker_socket_query()
juin 20 23:32:48 aaaa zabbix_agentd[6747]: Docker's socket query: GET /containers/ffff/stats HTTP/1.0

juin 20 23:32:18 aaaa zabbix_agentd[6746]: __zbx_zbx_setproctitle() title:'listener #2 [processing request]'
juin 20 23:32:18 aaaa zabbix_agentd[6746]: Requested [docker.stats[/ffff,memory_stats,usage]]
juin 20 23:32:18 aaaa zabbix_agentd[6746]: In zbx_module_docker_stats()
juin 20 23:32:18 aaaa zabbix_agentd[6746]: In zbx_module_docker_socket_query()
juin 20 23:32:18 aaaa zabbix_agentd[6746]: Docker's socket query: GET /containers/ffff/stats HTTP/1.0


juin 20 23:35:48 aaaa zabbix_agentd[8196]: __zbx_zbx_setproctitle() title:'listener #2 [processing request]'
juin 20 23:35:48 aaaa zabbix_agentd[8196]: Requested [docker.stats[/ffff,memory_stats,usage]]
juin 20 23:35:48 aaaa zabbix_agentd[8196]: In zbx_module_docker_stats()
juin 20 23:35:48 aaaa zabbix_agentd[8196]: In zbx_module_docker_socket_query()
juin 20 23:35:48 aaaa zabbix_agentd[8196]: Docker's socket query: GET /containers/ffff/stats HTTP/1.0

juin 20 23:37:03 aaaa zabbix_agentd[8195]: Metric source file: /sys/fs/cgroup/memory/docker/f49261aea5c7135b5e7ff046d2606836671c40f4dca31a28e106a8cbdd78c01e/memory.stat
juin 20 23:37:03 aaaa zabbix_agentd[8195]: Looking metric total_rss in memory.stat file
juin 20 23:37:03 aaaa zabbix_agentd[8195]: Id: f49261aea5c7135b5e7ff046d2606836671c40f4dca31a28e106a8cbdd78c01e; metric: total_rss; value: 1294745600
juin 20 23:37:03 aaaa zabbix_agentd[8195]: Sending back [9884680192]
juin 20 23:37:03 aaaa zabbix_agentd[8195]: __zbx_zbx_setproctitle() title:'listener #1 [waiting for connection]'
juin 20 23:37:03 aaaa zabbix_agentd[8195]: __zbx_zbx_setproctitle() title:'listener #1 [processing request]'
juin 20 23:37:03 aaaa zabbix_agentd[8195]: Requested [docker.stats[/ffff,memory_stats,usage]]
juin 20 23:37:03 aaaa zabbix_agentd[8195]: In zbx_module_docker_stats()
juin 20 23:37:03 aaaa zabbix_agentd[8195]: In zbx_module_docker_socket_query()
juin 20 23:37:03 aaaa zabbix_agentd[8195]: Docker's socket query: GET /containers/ffff/stats HTTP/1.0


juin 20 23:13:36 rrrr zabbix_agentd[451]: Cannot open metric file: '/sys/fs/cgroup/cpu,cpuacct/docker/6705a06097c4311144edfab00f32fb320a91232e141f0f40f3c1bcb78c85d108/cpuacct.stat'
juin 20 23:13:36 rrrr zabbix_agentd[450]: Cannot open metric file: '/sys/fs/cgroup/memory/docker/6705a06097c4311144edfab00f32fb320a91232e141f0f40f3c1bcb78c85d108/memory.stat'
juin 20 23:13:36 rrrr zabbix_agentd[451]: Cannot open metric file: '/sys/fs/cgroup/memory/docker/6705a06097c4311144edfab00f32fb320a91232e141f0f40f3c1bcb78c85d108/memory.stat'
juin 20 23:13:36 rrrr zabbix_agentd[451]: Cannot open metric file: '/sys/fs/cgroup/cpu,cpuacct/docker/6705a06097c4311144edfab00f32fb320a91232e141f0f40f3c1bcb78c85d108/cpuacct.stat'
juin 20 23:15:16 rrrr zabbix_agentd[450]: Cannot open metric file: '/sys/fs/cgroup/memory/docker//r-zabbix-agent-ct03-dca-1-ac96fcff/memory.stat'
juin 20 23:15:18 rrrr zabbix_agentd[450]: Cannot open metric file: '/sys/fs/cgroup/cpu,cpuacct/docker//r-zabbix-agent-ct03-dca-1-ac96fcff/cpuacct.stat'
juin 20 23:15:19 rrrr zabbix_agentd[450]: Cannot open metric file: '/sys/fs/cgroup/memory/docker//r-zabbix-agent-ct03-dca-1-ac96fcff/memory.stat'
juin 20 23:15:21 rrrr zabbix_agentd[450]: Cannot open metric file: '/sys/fs/cgroup/cpu,cpuacct/docker//r-zabbix-agent-ct03-dca-1-ac96fcff/cpuacct.stat'
juin 20 23:28:18 rrrr zabbix_agentd[3546]: Cannot open metric file: '/sys/fs/cgroup/cpu,cpuacct/docker/dbbbfac6b6e41ebe180fbf3baa2309de5114c9d85c287d1296dab78a289e3a16/cpuacct.stat'
juin 20 23:28:18 rrrr zabbix_agentd[3546]: Cannot open metric file: '/sys/fs/cgroup/cpu,cpuacct/docker/dbbbfac6b6e41ebe180fbf3baa2309de5114c9d85c287d1296dab78a289e3a16/cpuacct.stat'
juin 20 23:28:18 rrrr zabbix_agentd[3546]: Cannot open metric file: '/sys/fs/cgroup/memory/docker/dbbbfac6b6e41ebe180fbf3baa2309de5114c9d85c287d1296dab78a289e3a16/memory.stat'
juin 20 23:28:19 rrrr zabbix_agentd[3545]: Cannot open metric file: '/sys/fs/cgroup/memory/docker/dbbbfac6b6e41ebe180fbf3baa2309de5114c9d85c287d1296dab78a289e3a16/memory.stat'
juin 20 23:28:29 rrrr zabbix_agentd[3545]: Cannot open metric file: '/sys/fs/cgroup/memory/docker//r-zabbix-agent-ct03-dca-1-ac96fcff/memory.stat'
juin 20 23:28:30 rrrr zabbix_agentd[3545]: Cannot open metric file: '/sys/fs/cgroup/memory/docker//r-zabbix-agent-ct03-dca-1-ac96fcff/memory.stat'
juin 20 23:28:30 rrrr zabbix_agentd[3545]: Cannot open metric file: '/sys/fs/cgroup/cpu,cpuacct/docker//r-zabbix-agent-ct03-dca-1-ac96fcff/cpuacct.stat'
juin 20 23:28:30 rrrr zabbix_agentd[3545]: Cannot open metric file: '/sys/fs/cgroup/cpu,cpuacct/docker//r-zabbix-agent-ct03-dca-1-ac96fcff/cpuacct.stat'
juin 20 23:31:37 rrrr zabbix_agentd[6745]: Cannot open metric file: '/sys/fs/cgroup/memory/docker//r-zabbix-agent-ct03-dca-1-ac96fcff/memory.stat'
juin 20 23:31:37 rrrr zabbix_agentd[6746]: Cannot open metric file: '/sys/fs/cgroup/cpu,cpuacct/docker//r-zabbix-agent-ct03-dca-1-ac96fcff/cpuacct.stat', container doesn't run
juin 20 23:32:18 rrrr zabbix_agentd[6747]: Cannot open metric file: '/sys/fs/cgroup/cpu,cpuacct/docker/dbbbfac6b6e41ebe180fbf3baa2309de5114c9d85c287d1296dab78a289e3a16/cpuacct.stat'
juin 20 23:32:18 rrrr zabbix_agentd[6745]: Cannot open metric file: '/sys/fs/cgroup/memory/docker/dbbbfac6b6e41ebe180fbf3baa2309de5114c9d85c287d1296dab78a289e3a16/memory.stat'
juin 20 23:32:48 rrrr zabbix_agentd[6747]: Cannot open metric file: '/sys/fs/cgroup/cpu,cpuacct/docker/dbbbfac6b6e41ebe180fbf3baa2309de5114c9d85c287d1296dab78a289e3a16/cpuacct.stat', container doesn't run
juin 20 23:35:48 rrrr zabbix_agentd[8197]: Cannot open metric file: '/sys/fs/cgroup/memory/docker/dbbbfac6b6e41ebe180fbf3baa2309de5114c9d85c287d1296dab78a289e3a16/memory.stat'
juin 20 23:36:07 rrrr zabbix_agentd[8195]: Cannot open metric file: '/sys/fs/cgroup/cpu,cpuacct/docker//r-zabbix-agent-ct03-dca-1-ac96fcff/cpuacct.stat'
juin 20 23:36:07 rrrr zabbix_agentd[8197]: Cannot open metric file: '/sys/fs/cgroup/cpu,cpuacct/docker//r-zabbix-agent-ct03-dca-1-ac96fcff/cpuacct.stat', container doesn't run
juin 20 23:40:52 rrrr zabbix_agentd[11699]: Cannot open metric file: '/sys/fs/cgroup/memory/docker/dbbbfac6b6e41ebe180fbf3baa2309de5114c9d85c287d1296dab78a289e3a16/memory.stat'
juin 20 23:40:52 rrrr zabbix_agentd[11699]: Cannot open metric file: '/sys/fs/cgroup/cpu,cpuacct/docker//r-zabbix-agent-ct03-dca-1-ac96fcff/cpuacct.stat', container doesn't run
juin 20 23:40:52 rrrr zabbix_agentd[11698]: Cannot open metric file: '/sys/fs/cgroup/cpu,cpuacct/docker/dbbbfac6b6e41ebe180fbf3baa2309de5114c9d85c287d1296dab78a289e3a16/cpuacct.stat', container doesn't run
juin 20 23:41:23 rrrr zabbix_agentd[11698]: Cannot open metric file: '/sys/fs/cgroup/cpu,cpuacct/docker//r-zabbix-agent-ct03-dca-1-ac96fcff/cpuacct.stat', container doesn't run
juin 20 23:45:16 rrrr zabbix_agentd[14744]: Cannot open metric file: '/sys/fs/cgroup/memory/docker//r-zabbix-agent-ct03-dca-1-ac96fcff/memory.stat'
juin 20 23:45:17 rrrr zabbix_agentd[14744]: Cannot open metric file: '/sys/fs/cgroup/cpu,cpuacct/docker//r-zabbix-agent-ct03-dca-1-ac96fcff/cpuacct.stat'
```

**Comment by ptempier** [comment]
The last few crashes i traced are from a docker which is in started once mode in rancher and thus have no memory stats or anything

**Comment by jangaraj** [comment]
What's your zabbix_agent `Timeout` config? Module uses this timeout value in `zbx_module_item_timeout` function.

**Comment by ptempier** [comment]
it s default value ( i think it s  3 seconds)

**Comment by jangaraj** [comment]
OK, what about docker socket. Is it available?:
```
echo -e "GET /containers/json?all=1 HTTP/1.0\r\n" | nc -U /var/run/docker.sock
HTTP/1.0 200 OK
Api-Version: 1.26
Content-Type: application/json
Docker-Experimental: false
Server: Docker/17.03.0-ce (linux)
Date: Wed, 21 Jun 2017 07:06:50 GMT

[{"Id":"d526bed637f8fb7fc3993....
```

**Comment by ptempier** [comment]
it works with :  su -s /bin/bash -c  /tmp/aa.sh zabbix
aa.sh is : echo -e "GET /containers/json?all=1 HTTP/1.0\r\n" | nc -U /var/run/docker.sock

**Comment by jangaraj** [comment]
OK, And what about that problematic command?
```
echo -e "GET /containers/ffff/stats HTTP/1.0\r\n" | nc -U /var/run/docker.sock
```

**Comment by ptempier** [comment]
the command get stuck on :

 su -s /bin/bash -c  /var/docker/tmp/aa.sh zabbix
HTTP/1.0 200 OK
Server: Docker/1.12.6 (linux)
Date: Thu, 22 Jun 2017 08:34:56 GMT
Content-Type: text/plain; charset=utf-8

**Comment by jangaraj** [comment]
And what about docker stats?
```
docker stats ffff
```

**Comment by ptempier** [comment]
Hello

Its  empty, container is not running

CONTAINER         CPU %               MEM USAGE / LIMIT   MEM %        NET I/O             BLOCK I/O           PIDS
fffffffff   --                  -- / --             --                  -- / --             -- / --             --

**Comment by jangaraj** [comment]
What's your docker config and `docker info`?

**Comment by ptempier** [comment]
Containers: 80
 Running: 63
 Paused: 0
 Stopped: 17
Images: 70
Server Version: 1.12.6
Storage Driver: overlay2
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: null overlay host bridge
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: 
containerd version: 
runc version: 
init version: 
Security Options:
 seccomp
Kernel Version: 4.9.11-1-ARCH
Operating System: Arch Linux
OSType: linux
Architecture: x86_64
CPUs: 88
Total Memory: 503.8GiB
Name: yyyy
ID: W4TB:xxxx
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

**Comment by jangaraj** [comment]
I don't see any obvious problem. That Docker stat request should fail immediately:
```
curl --unix-socket /var/run/docker.sock -X GET http:/v1.24/containers/ffff/stats
{"message":"No such container: ffff"}
```
But it gets stuck. I'll try to implement docker socket query timeout, so that docker query will timeout after 1-30 sec (specified in the agent `Timeout` setting).

**Comment by ptempier** [comment]
thanks

**Comment by jangaraj** [comment]
Could you test branch `cstatus` please? Unfortunately I'm not able to test it, because I don't have any docker daemon with your behavior. Docker socket connection should reach timeout, because socket option `SO_SNDTIMEO` is set (value is from the zabbix_agent `Timeout` config). 

Doc: https://www.freebsd.org/cgi/man.cgi?query=setsockopt&sektion=2
```
SO_SNDTIMEO is an option to set a timeout value for output	operations.
     It	accepts	a struct timeval argument with the number of seconds and
     microseconds used to limit	waits for output operations to complete.  If a
     send operation has	blocked	for this much time, it returns with a partial
     count or with the error EWOULDBLOCK if no data were sent.	In the current
     implementation, this timer	is restarted each time additional data are
     delivered to the protocol,	implying that the limit	applies	to output por-
     tions ranging in size from	the low	water mark to the high water mark for
     output.
```

**Comment by ptempier** [comment]
Hello

So far so good, no crash since yesterday (before it was crashing after a few minutes)

**Comment by jangaraj** [comment]
Fixed in the v0.6.7 release - https://github.com/monitoringartist/zabbix-docker-monitoring/pull/76
